### ISSUE TITLE
Where were the style transfer example models taken from?

### ISSUE THREAD

**Issue by AndreJFBico** [xmartlabs/Bender#49]
Hi, if possible, do you guys know from what example the q_and_w.pb and g_and_w2.pb files were taken from. Im trying to get similar style transfer models to work with bender but the list of unsupported operations varies widely.

If you trained them yourself it would be helpful to know more information about those models.

PS:i would have posted this on stack-overflow but i don't have the rep to create the bender tag.
Thanks in advance!

**Comment by bryant1410** [comment]
I think those protobufs were generated from [fast-style-transfer](https://github.com/lengstrom/fast-style-transfer).

Just out of curiosity, which are the unsupported operations that you ran into? Also, remember that you can try implementing them yourself. It's not that hard :sweat_smile: We can provide help :angel: 

I don't have the enough rep to create a tag either :disappointed: But you can go ahead and ask without the tag. Hopefully, the community will create it some day.

**Comment by AndreJFBico** [comment]
Hey ok ill try it and if i run into issues ill post them, ill close this issue in the meantime.

Thanks!

**Comment by mdramos** [comment]
@bryant1410 would you be willing to share how you generated the protobufs from fast-style-transfer? 

I've created a model using the same library, but that only generates checkpoint files and no graphs

**Comment by bryant1410** [comment]
I don't have them at hand now. I suggest you use https://github.com/xmartlabs/benderthon in conjunction with fast-style-transfer. If you are blocked, please tell me and maybe I can help you out.

**Comment by mdramos** [comment]
Ok thanks, my only blocker right now is generating the pb file. Currently Im using the approach in this SO answer [here](https://stackoverflow.com/questions/45864363/tensorflow-how-to-convert-meta-data-and-index-model-files-into-one-graph-pb)

But need to see how I can generate the output node names in fast-style-transfer

**Comment by mdramos** [comment]
I guess my next approach would be to use Benderthon, but would need to create the single checkpoint file, as I currently have the 3 (meta, index, data)

**Comment by bryant1410** [comment]
> I guess my next approach would be to use Benderthon, but would need to create the single checkpoint file, as I currently have the 3 (meta, index, data)

You don't need a single one, you can with 3. You just need to pass the common prefix (like "file.ckpt")

**Comment by bryant1410** [comment]
And with respect to the output node name, you just need to check its name

**Comment by mdramos** [comment]
Do you recall what it was for fast-style-transfer, it doesnt seem to be explicitly set: https://github.com/lengstrom/fast-style-transfer/search?utf8=%E2%9C%93&q=name%3D&type=

**Comment by mdramos** [comment]
```
saver = tf.train.import_meta_graph(/path/to/meta/graph)
sess = tf.Session()
saver.resore(sess, /path/to/checkpoints)
graph = sess.graph
print([node.name for node in graph.as_graph_def().node])
```

produces a ton of different output

**Comment by bryant1410** [comment]
> saver = tf.train.import_meta_graph(/path/to/meta/graph)
> sess = tf.Session()
> saver.resore(sess, /path/to/checkpoints)
> graph = sess.graph
> print([node.name for node in graph.as_graph_def().node])
> produces a ton of different output

That yields every node name in the graph, not necessarily output.

**Comment by bryant1410** [comment]
The output is in: https://github.com/lengstrom/fast-style-transfer/blob/master/src/transform.py#L18

It's automatically named by TF, probably by the Add operation, something like "AddN" (where N is the number of the last unnamed addition operation, 0-based).

**Comment by mdramos** [comment]
ah thanks, looks like I was able to find it in the evaluate function and the restore of the graph:
`Tensor("add_37:0", shape=(1, 720, 884, 3), dtype=float32, device=/device:GPU:0)`

now when running `benderthon tf-freeze fns.ckpt tiger.pb add_37:0` I get error:
`AttributeError: 'module' object has no attribute 'freeze_graph_with_def_protos'`

**Comment by bryant1410** [comment]
can you pass a complete stacktrace?

**Comment by bryant1410** [comment]
what version of tf are you using?

**Comment by mdramos** [comment]
**version 1.1.0**

```
Traceback (most recent call last):
  File "/home/ubuntu/anaconda2/bin/benderthon", line 11, in <module>
    sys.exit(main())
  File "/home/ubuntu/anaconda2/lib/python2.7/site-packages/benderthon/cmdline.py", line 39, in main
    tf_freeze.freeze_from_checkpoint(args.input_checkpoint, args.output_path, args.output_node_names.split(','))
  File "/home/ubuntu/anaconda2/lib/python2.7/site-packages/benderthon/tf_freeze.py", line 25, in freeze_from_checkpoint
    freeze_graph.freeze_graph_with_def_protos(input_graph_def=sess.graph_def, input_saver_def=None,
AttributeError: 'module' object has no attribute 'freeze_graph_with_def_protos'
```

**Comment by mdramos** [comment]
sorry just saw the 1.2 req

**Comment by mdramos** [comment]
Thanks again. I was able to get the pb file.

So far in my iOS style-transfer I can only get `g_and_w2.pb` (given) to work. Seeing some `nil` errors with my transfered pb's. Will update when I find out more

**Comment by mdramos** [comment]
ah the `nil` error was my b.

But the newest error is one I believe I saw in another issue:
`assertion failed: Add works for two layers:`

**Comment by j005u** [comment]
Having the same issue
`assertion failed: Add works for two layers:`
Here's my protobuf: https://www.dropbox.com/s/k4i8ccn0ak0kac9/c.pb?dl=0
Generated with benderthon from a checkpoint trained with https://github.com/lengstrom/fast-style-transfer

This is NOT the same as #50.

**Comment by bryant1410** [comment]
@j005u why is not the same issue?

**Comment by j005u** [comment]
Because the error is `Add works for two layers` rather than `Add works for two layers of the same size`

More specifically incoming.count = 0

**Comment by j005u** [comment]
@bryant1410 Uploaded the original checkpoint to https://www.dropbox.com/s/h2ck8xf8h576kvx/check.ckpt.zip?dl=0

The output layer is add_37.

Also of note (although I'm guessing it doesn't matter?), the model was originally trained for cloud GPUs at 1980x1080. 

Maybe that helps.

**Comment by j005u** [comment]
@bryant1410 Any tips on debugging this?

I've tried old version of benderthon and Tensorflow. I also tried quantising the graph, but that just results in a conv2d layer having 0 inputs. Could the trained input size be an issue, should I try to retrain at 256x256? Is there a tool I could use to visually explore the .pb and try to figure out where things break down? As far as I can tell the very first Add layer has 0 inputs. 

@mdramos Did you ever figure this out by any chance?

As far as I can tell there's no way to use fast-style-transfer with Bender currently.
